### ISSUE TITLE
TorchDynamo + AoTAutograd Graph behavior with Softmax causing slowdown

### ISSUE THREAD

**Issue by parthmannan** [pytorch/pytorch#86197]
### 🐛 Describe the bug

Hi, I am running a small sample of dot-product attention (Scale-Mask-Softmax-Dropout) with TorchDynamo + AoTAutograd enabled.

Here's a code snippet to reproduce the behavior that's described below the code.
```
import torch
import torch.nn as nn
import torch.nn.functional as F
import argparse
import torchdynamo

tensor_dtype = torch.float16

ap = argparse.ArgumentParser()
ap.add_argument("-td", "--torchdynamo", action='store_true', default=False, help='Enable Nvfuser + TorchDynamo optimization pass')

args = vars(ap.parse_args())
torchdynamo_enable = bool(args['torchdynamo'])

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.scale = 8
        self.mask = torch.tril(torch.ones([16, 1, 2048, 2048], dtype=torch.float16, device='cuda'))
        self.softmax = nn.Softmax(dim=-1)

    def forward(self, attention_scores):
        #Scale-Mask-Softmax-Dropout
        out = attention_scores * self.scale
        out = out + ((1.0 - self.mask) * -10000.0)
        #out = F.gelu(out)
        out = self.softmax(out)
        out = F.dropout(out, p=0.1, training=True)

        return out

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

net = Net()
net = net.cuda()
net = net.half()

if torchdynamo_enable:
    network_fn = torchdynamo.optimize("aot_nvfuser")(net)
else:
    network_fn = torch.jit.script(net)

shape = [16, 16, 2048, 2048]
q = torch.rand(shape, dtype=tensor_dtype, requires_grad=True, device=device)
t = torch.rand(shape, dtype=tensor_dtype, requires_grad=True, device=device)

outputs = network_fn(q)
outputs.backward(t)
```

One can use the argument `--torchdynamo` to enable TorchDynamo + AoTAutograd. Default is using TorchScript.

This is the default graph for the code above - 

![default_graph_attn](https://user-images.githubusercontent.com/38387286/193737767-8e9be67f-be70-4bab-bc7c-4c412d61fd92.PNG)

When enabling TorchDynamo + AoTAutograd, the algo chooses to recompute parts of forward pass in the backward pass to save memory writes in forward. 

![aot_graph_attn](https://user-images.githubusercontent.com/38387286/193747379-a884c1ba-72e9-463c-9ee4-ee22b78fe878.PNG)

Attached the graph output with `AOT_FX_GRAPHS=1`
[aot_dynamo_mha.log](https://github.com/pytorch/pytorch/files/9703323/aot_dynamo_mha.log)

Although, this may make for a good choice given bandwidth savings as the mask is cheaper than the softmax fprop output but the overall time is much slower than using TorchScript. This is potentially due to recomputing the expensive Softmax operation in backward.

Another potential issue is that the algo decides to write out the output of the sub op which seems unnecessary given that the input to sub (mask) is already in memory and this sub can also be recomputed as well. To summarize, I believe there are two issues - 

a) The choice to recompute Softmax is probably too expensive.
b) The choice where the algo makes the cut can potentially be improved to further save memory traffic.

Thanks.

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0a0+git4bdc0af
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-156-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.8.86
GPU models and configuration: GPU 0: NVIDIA A100 80GB PCIe
Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.4.0
[pip3] CoCa-pytorch==0.0.6
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.0.10
[pip3] functorch==0.3.0a0+4bdc0af
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.0
[pip3] rotary-embedding-torch==0.1.5
[pip3] torch==1.13.0a0+git4bdc0af
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchdynamo==1.13.0.dev0
[pip3] torchfile==0.1.0
[pip3] torchmetrics==0.9.3
[pip3] torchrec-nightly==2022.9.13
[pip3] torchtext==0.14.0a0+67d2692
[pip3] torchvision==0.14.0a0+6945201
[pip3] torchx-nightly==2022.9.15
[pip3] vector-quantize-pytorch==0.9.2
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.4.0                    pypi_0    pypi
[conda] coca-pytorch              0.0.6                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.0.10                   pypi_0    pypi
[conda] functorch                 0.3.0a0+4bdc0af           dev_0    <develop>
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] nomkl                     3.0                           0  
[conda] numpy                     1.21.2                   pypi_0    pypi
[conda] numpy-base                1.21.5           py38hb8be1f0_2  
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.0                    pypi_0    pypi
[conda] rotary-embedding-torch    0.1.5                    pypi_0    pypi
[conda] torch                     1.13.0a0+git4bdc0af           dev_0    <develop>
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchdynamo               1.13.0.dev0               dev_0    <develop>
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchrec-nightly          2022.9.13                pypi_0    pypi
[conda] torchtext                 0.14.0a0+67d2692           dev_0    <develop>
[conda] torchvision               0.14.0a0+6945201           dev_0    <develop>
[conda] torchx-nightly            2022.9.15                pypi_0    pypi
[conda] vector-quantize-pytorch   0.9.2                    pypi_0    pypi

```

cc @zou3519 @Chillee @samdow @soumith

**Comment by IvanYashchuk** [comment]
It seems possible to ban softmax from recomputation by modifying this list:
https://github.com/pytorch/pytorch/blob/736adc08084814c13d57324f74adad091e304eb2/functorch/_src/partitioners.py#L278
or this function
https://github.com/pytorch/pytorch/blob/736adc08084814c13d57324f74adad091e304eb2/functorch/_src/partitioners.py#L309

**Comment by Chillee** [comment]
@IvanYashchuk Ah, good catch.

I think @IvanYashchuk is right - this is just an oversight/mistake. We should remove `softmax` from the list.

Generally, we do not allow recomputing normalizations (i.e. `native_layer_norm` and such), I think softmax was added as a mistake.

**Comment by Chillee** [comment]
I verified @IvanYashchuk's suggestion:

```
import torch.nn as nn

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.scale = 8
        self.mask = torch.tril(torch.ones([16, 1, 2048, 2048], dtype=torch.float16, device='cuda'))
        self.softmax = nn.Softmax(dim=-1)

    def forward(self, attention_scores):
        #Scale-Mask-Softmax-Dropout
        out = attention_scores * self.scale
        out = out + ((1.0 - self.mask) * -10000.0)
        #out = F.gelu(out)
        out = self.softmax(out)
        out = F.dropout(out, p=0.1, training=True)

        return out

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
tensor_dtype = torch.float16

net = Net()
net = net.cuda()
net = net.half()

network_fn = torchdynamo.optimize("aot_nvfuser")(net)

shape = [16, 16, 2048, 2048]
q = torch.rand(shape, dtype=tensor_dtype, requires_grad=True, device=device)
t = torch.rand(shape, dtype=tensor_dtype, requires_grad=True, device=device)

def bench(f):
    for _ in range(5):
        f()
    torch.cuda.synchronize()
    begin = time.time()
    for _ in range(100):
        f()
    torch.cuda.synchronize()
    print((time.time()-begin)*1e6/100)

bench(lambda: network_fn(q).backward(t))
```

```
Before: 26246.845722198486
After: 21274.752616882324
```

**Comment by Chillee** [comment]
Fixed by https://github.com/pytorch/pytorch/pull/86268

**Comment by Chillee** [comment]
As for the second issue, that's another good catch - `aten.rsub` also isn't in the list of "recomputable" ops today.

**Comment by Chillee** [comment]
I think this should be resolved now?
